### PR TITLE
Add `Channel` worker

### PR DESCRIPTION
**Pull request by cezarad**
Closes: #822

## Description

- added a worker for  channel 
- added query_connection_channels to the chain handle and runtime
- redo #923 to return `Uninitialized` when empty value bytes are returned from the ABCI query for connection and channel. Changed the CLI to return error if connection/ channel does not exist.
- allow entering the destination channel id in `hermes tx raw chan-open-try...` and make `dst_channel_id` and option (used to be `ChannelId::default()` when not specified but this is a valid channel id.
- change `strategy` in config to be either `packets`(only relay packets) or `all` (relay from all events, including channel handshake ones)
- e2e CI changes:
   - added channel worker tests (see below)
   - determine the chain names from config file
   - return the parameters from the test (`raw`) that creates the clients, connection, channel for the next `start` test (instead of hardcoding to `connection-0`, `channel-0` etc). This way we can run `e2e/run.py` multiple times locally without restarting the chains.

To setup for the tests:
- change `strategy` in config to `all` (relay from all events, including channel handshake ones)
- execute start the chains using  `./scripts/dev-env ~/.hermes/config.toml ibc-0 ibc-1` and then 
- create a connection using create a connection : ` hermes create connection ibc-0 ibc-1 `

Here are a few tests (also added to `e2e` CI):
- case A: 
      1. start the relayer ` hermes start`    
      2. send `hermes tx raw chan-open-init ibc-0 ibc-1 connection-0 transfer transfer `
- Case B: 
     1. send `hermes tx raw chan-open-init ibc-0 ibc-1 connection-0 transfer transfer `
     2. start the relayer ` hermes start` 
- Case C: 
     1. send `hermes tx raw chan-open-init ibc-0 ibc-1 connection-0 transfer transfer `
     2. send `hermes tx raw chan-open-try ibc-1 ibc-0 connection-0 transfer transfer -s channel-x `, with `channel-x` being the ID of the channel returned in step 1 above.
     3. start the relayer ` hermes start ` 
- Case D: 
     1. send `hermes tx raw chan-open-init ibc-0 ibc-1 connection-0 transfer transfer `
     2. send `hermes tx raw chan-open-try ibc-1 ibc-0 connection-0 transfer transfer -s channel-x `, with `channel-x` being the ID of the channel returned in step 1 above
     3. send `hermes tx raw chan-open-ack ibc-0 ibc-1 connection-0 transfer transfer -d channel-x -s channel-y `, with `channel-x` being the ID of the channel returned in step 1 and `channel-y` the ID of the channel returned in step 2 above
     4. start the relayer ` hermes start ` 

After the last step, wait a couple of seconds for hermes to finish the handshake and verify the channel state on both chains using `hermes query channel end ibc-0 transfer channel-x` `hermes query channel end ibc-0 transfer channel-y`. They should both be in open state.

- TODO before merge:
  - update guide or mark issue with proper tag
  - change CI to `packets` strategy. Will add a new CI to use a config with `all` later
______

For contributor use:

- [x] Updated the __Unreleased__ section of [CHANGELOG.md](https://github.com/informalsystems/ibc-rs/blob/master/CHANGELOG.md) with the issue.
- [x] If applicable: Unit tests written, added test to CI.
- [x] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [ ] Updated relevant documentation (`docs/`) and code comments.
- [ ] Re-reviewed `Files changed` in the Github PR explorer.

### CODE REVIEW

**Comment by ancazamfir** [comment]
Changed to draft to cleanup my changes

**Review by adizere — APPROVED** [review]
Massive work Anca & Cezara!

Had a look over the code and looks good, the tests also run fine. I am still running some further tests with `gm` but I'm approving already since I consider this done. Mind the "TODO before merge" remark.

File: docs/architecture/adr-002-ibc-relayer.md [hunk 1]
```diff
@@ -200,7 +200,7 @@ pub struct Config {
 }
 
 pub enum Strategy {
-    Naive,
+    Packets,
```
Comment (romac) [inline docs/architecture/adr-002-ibc-relayer.md:203]:
```suggestion
    All,
    Packets,
```

File: relayer/src/worker/channel.rs [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+                        let last_event = batch.events.last();
+                        debug!("channel worker starts processing {:#?}", last_event);
+                        match last_event {
+                            Some(event) => {
+                                handshake_channel = RelayChannel::restore_from_event(
+                                    a_chain.clone(),
+                                    b_chain.clone(),
+                                    event.clone(),
+                                )?;
+                                retry_with_index(retry_strategy::uni_chan_path(), |index| {
```
[the remark this answers is not in the document]
Comment (ancazamfir) [reply relayer/src/worker/channel.rs:58]: renamed to `worker_default_strategy`

File: relayer/src/supervisor.rs [hunk 3]
```diff
@@ -78,20 +81,68 @@ impl Supervisor {
                         }
                     }
                 }
+
+                IbcEvent::OpenInitChannel(ref _open_init) => {
+                    if !self.handshake_enabled() {
+                        continue;
+                    }
+                    if let Ok(object) = Object::channel_from_chan_open_events(
+                        event.clone().channel_attributes().unwrap(),
```
[the remark this answers is not in the document]
Comment (ancazamfir) [reply relayer/src/supervisor.rs:90]: missed this one

File: relayer/src/channel.rs [hunk 4]
```diff
@@ -255,14 +389,23 @@ impl Channel {
         while counter < MAX_RETRIES {
             counter += 1;
 
+            assert!(self.src_channel_id().is_some());
+            assert!(self.dst_channel_id().is_some());
```
[the remark this answers is not in the document]
Comment (ancazamfir) [reply relayer/src/channel.rs:393]: missed this also, thanks!

File: relayer/src/channel.rs [hunk 5]
```diff
@@ -141,6 +155,126 @@ impl Channel {
...
+            ),
+            b_side: ChannelSide::new(
+                counterparty_chain.clone(),
+                connection.counterparty().client_id().clone(),
+                counterparty_connection_id.clone(),
+                channel_event_attributes.counterparty_port_id.clone(),
+                channel_event_attributes.counterparty_channel_id.clone(),
+            ),
+            connection_delay: connection.delay_period(),
+            //TODO  detect version from event
```
[the remark this answers is not in the document]
Comment (ancazamfir) [reply relayer/src/channel.rs]: Removed and added explanation. We cannot do much here but the message builders figure it out.

File: relayer/src/channel.rs [hunk 6]
```diff
@@ -141,6 +155,126 @@ impl Channel {
...
+        let connection = chain.query_connection(&connection_id, Height::zero())?;
+        let connection_counterparty = connection.counterparty();
+
+        let counterparty_connection_id = connection_counterparty
+            .connection_id()
+            .ok_or(ChannelError::MissingCounterpartyConnection)?;
+
+        Ok(Channel {
+            ordering: Default::default(),
+            //TODO  how to get the order from event
```
[the remark this answers is not in the document]
Comment (ancazamfir) [reply relayer/src/channel.rs]: Removed and added explanation. We cannot do much here but the message builders figure it out.

**Review by romac — APPROVED** [review]

**Review by ancazamfir — CHANGES_REQUESTED** [review]
change strategy in CI

**Pull request merged**